File: 단계별로 풀어보기/Step 16/2579.cpp
#include <iostream>

using namespace std;

int main() {
    int n;
    int stairs[301];
    int score[301];
    int i;

    cin >> n;
    for(i=1; i<=n; i++) {
        cin >> stairs[i];
    }

    score[1] = stairs[1];
    score[2] = max(stairs[2], score[1]+stairs[2]);
    score[3] = max(score[1]+stairs[3], stairs[2]+stairs[3]);

    for(i=4; i<=n; i++) {
        score[i] = max(score[i-2]+stairs[i], score[i-3]+stairs[i-1]+stairs[i]);
    }
    cout << score[n] << endl;
    return 0;
}

File: StepByStep/Python/Step_16_동적 계획법 1/2579.py
import sys
N = int(input())
score = [0] * 301
dp_score = [0] * 301

for s in range(N):
    score[s] = int(sys.stdin.readline())

dp_score[0] = score[0]
dp_score[1] = score[0]+score[1]
dp_score[2] = max(score[0]+score[2], score[1]+score[2])

for i in range(3, N):
    dp_score[i] = max(dp_score[i-3]+score[i-1]+score[i], dp_score[i-2]+score[i])

print(dp_score[N-1])

